srv/modules/runners/ext_lib/utils.py

```diff
@@ -85,6 +87,157 @@ def evaluate_state_return(job_data):
+            log_n_print(f"role-{role_name} is *NOT* running on {minion}")
+            running = False
+    return running
+
+
+def _remove_role(role=None, non_interactive=False):
+    # TODO: already_running vs is_running
+    # find process id vs. systemd
+    ##
+    ## There is mon ok-to-rm, ok-to-stop, ok-to-add-offline
+    ##
+    """ TODO: docstring """
+    assert role
+    already_running = LocalClient().cmd(
+        f"I@cluster:ceph and not I@roles:{role}",
+        f'{role}.already_running',
+        tgt_type='compound')
+    to_remove = [k for (k, v) in already_running.items() if v]
+    if not to_remove:
+        print("Nothing to remove. Exiting..")
+        return True
+    if prompt(
+            f"""Removing role: {role} on minion {', '.join(to_remove)}
+Continue?""",
+            non_interactive=non_interactive,
+            default_answer=True):
+        print(f"Removing {role} on {' '.join(to_remove)}")
+        ret: str = LocalClient().cmd(
+            to_remove,
+            f'podman.remove_{role}',
+            ['registry.suse.de/devel/storage/6.0/images/ses/6/ceph/ceph'],
+            tgt_type='list')
+        if not evaluate_module_return(ret):
+            return False
+
+        ret = [is_running(minion, role_name=role, func='wait_role_down') for minion in to_remove]
+        # TODO: do proper checks here:
+        if all(ret):
+            print(f"{role} deletion was successful.")
+            return True
+        return False
+
+
+    else:
+        return 'aborted'
+
+
+def _deploy_role(role=None, non_interactive=False):
+    assert role
+    candidates = _get_candidates(role=role)
+    if candidates:
+        if prompt(
+                f"""These minions will be {role}: {', '.join(candidates)}
+Continue?""",
+                non_interactive=non_interactive,
+                default_answer=True):
+            print("Deploying..")
+
+            if role == 'mgr':
+                for candidate in candidates:
+                    # create and register keyring
+                    ret: str = LocalClient().cmd(
+                        "roles:master",
+                        f'podman.create_mgr_keyring',
+                        ['registry.suse.de/devel/storage/6.0/images/ses/6/ceph/ceph', candidate],
+                        tgt_type='pillar')
+
+
+                if not evaluate_module_return(ret):
+                    return False
+
+                # distrubute keyring
+                ret: str = LocalClient().cmd(
+                    "roles:mgr",
+                    'state.apply',
+                    ['ceph.mgr.keyring'],
+                    tgt_type='pillar')
+
+                if not evaluate_state_return(ret):
+                    return False
+
+
+            ret: str = LocalClient().cmd(
+                candidates,
+                f'podman.create_{role}',
+                ['registry.suse.de/devel/storage/6.0/images/ses/6/ceph/ceph'],
```

Comment:
How and where to set the ceph_container_image is not yet answered I guess. For this we need to figure out how the update/upgrade process works.

While I see the the need for a global const, I have to point out that we need this const in a cluster wide 'global' context. Speaking in salt terms, this would probably be the pillar/grain.

We should follow up on this topic soon-ish